File: tests/algorithms/x/Clojure/maths/largest_of_very_large_numbers.clj
(ns main (:refer-clojure :exclude [ln log10 absf res test_res compare]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare ln log10 absf res test_res compare)

(def ^:dynamic compare_r1 nil)

(def ^:dynamic compare_r2 nil)

(def ^:dynamic ln_k nil)

(def ^:dynamic ln_sum nil)

(def ^:dynamic ln_t nil)

(def ^:dynamic ln_term nil)

(defn ln [ln_x]
  (binding [ln_k nil ln_sum nil ln_t nil ln_term nil] (try (do (set! ln_t (/ (- ln_x 1.0) (+ ln_x 1.0))) (set! ln_term ln_t) (set! ln_sum 0.0) (set! ln_k 1) (while (<= ln_k 99) (do (set! ln_sum (+ ln_sum (/ ln_term (double ln_k)))) (set! ln_term (* (* ln_term ln_t) ln_t)) (set! ln_k (+ ln_k 2)))) (throw (ex-info "return" {:v (* 2.0 ln_sum)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn log10 [log10_x]
  (try (throw (ex-info "return" {:v (/ (ln log10_x) (ln 10.0))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn absf [absf_x]
  (try (if (< absf_x 0.0) (- absf_x) absf_x) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn res [res_x res_y]
  (try (do (when (= res_x 0) (throw (ex-info "return" {:v 0.0}))) (when (= res_y 0) (throw (ex-info "return" {:v 1.0}))) (when (< res_x 0) (throw (Exception. "math domain error"))) (throw (ex-info "return" {:v (* (double res_y) (log10 (double res_x)))}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn test_res []
  (do (when (> (absf (- (res 5 7) 4.892790030352132)) 0.0000001) (throw (Exception. "res(5,7) failed"))) (when (not= (res 0 5) 0.0) (throw (Exception. "res(0,5) failed"))) (when (not= (res 3 0) 1.0) (throw (Exception. "res(3,0) failed")))))

(defn compare [compare_x1 compare_y1 compare_x2 compare_y2]
  (binding [compare_r1 nil compare_r2 nil] (try (do (set! compare_r1 (res compare_x1 compare_y1)) (set! compare_r2 (res compare_x2 compare_y2)) (when (> compare_r1 compare_r2) (throw (ex-info "return" {:v (str (str (str "Largest number is " (str compare_x1)) " ^ ") (str compare_y1))}))) (if (> compare_r2 compare_r1) (str (str (str "Largest number is " (str compare_x2)) " ^ ") (str compare_y2)) "Both are equal")) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (test_res)
      (println (compare 5 7 4 8))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Java/maths/largest_of_very_large_numbers.java
public class Main {

    static double ln(double x) {
        double t = (x - 1.0) / (x + 1.0);
        double term_1 = t;
        double sum_1 = 0.0;
        long k_1 = 1;
        while (k_1 <= 99) {
            sum_1 = sum_1 + term_1 / (((Number)(k_1)).doubleValue());
            term_1 = term_1 * t * t;
            k_1 = k_1 + 2;
        }
        return 2.0 * sum_1;
    }

    static double log10(double x) {
        return ln(x) / ln(10.0);
    }

    static double absf(double x) {
        if (x < 0.0) {
            return -x;
        }
        return x;
    }

    static double res(long x, long y) {
        if (x == 0) {
            return 0.0;
        }
        if (y == 0) {
            return 1.0;
        }
        if (x < 0) {
            throw new RuntimeException(String.valueOf("math domain error"));
        }
        return (((Number)(y)).doubleValue()) * log10(((Number)(x)).doubleValue());
    }

    static void test_res() {
        if (absf(res(5, 7) - 4.892790030352132) > 1e-07) {
            throw new RuntimeException(String.valueOf("res(5,7) failed"));
        }
        if (res(0, 5) != 0.0) {
            throw new RuntimeException(String.valueOf("res(0,5) failed"));
        }
        if (res(3, 0) != 1.0) {
            throw new RuntimeException(String.valueOf("res(3,0) failed"));
        }
    }

    static String compare(long x1, long y1, long x2, long y2) {
        double r1 = res(x1, y1);
        double r2_1 = res(x2, y2);
        if (r1 > r2_1) {
            return "Largest number is " + _p(x1) + " ^ " + _p(y1);
        }
        if (r2_1 > r1) {
            return "Largest number is " + _p(x2) + " ^ " + _p(y2);
        }
        return "Both are equal";
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            test_res();
            System.out.println(compare(5, 7, 4, 8));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            if (d == Math.rint(d)) return String.valueOf((long) d);
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}
